### PR TITLE
leptonica: fix CMake imported target + honor options + add with_zlib option

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **leptonica/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Actually it's a big refactoring:
- delete fPIC option if shared
- move fix for 1.78.0 in a patch file
- `with_zlib` option is added since it exists in upstream autoconf file
- default value of `with_gif` is now True (looking at upstream build files, there was no reason that this option had a different default value than others options).
- a classic CMakeLists wrapper is used, it perfectly works with a tiny fix
- it's more robust to rely on pkgconf for openjpeg & libwebp dependencies
- the all mechanism of `CMAKE_DISABLE_FIND_PACKAGE_<LIB>` -used to honor options- can't work in this recipe, because `CMAKE_DISABLE_FIND_PACKAGE_<LIB>` is forbidden when find_package(LIB REQUIRED) is called. Actually leptonica doesn't use REQUIRED, but... find modules generated do, and `CMAKE_DISABLE_FIND_PACKAGE_<LIB>` is global ! Some dependencies of leptonica are also dependenies of its dependencies (libjpeg and zlib for example are dependencies of libtiff), therefoe it breaks. /cc @theirix
- fix CMake imported target which is `leptonica` (without namespace)

### CODE REVIEW

File: recipes/leptonica/all/conanfile.py [hunk 1]
```diff
@@ -1,42 +1,68 @@
...
+import textwrap
+
+required_conan_version = ">=1.33.0"
 
 
 class LeptonicaConan(ConanFile):
     name = "leptonica"
     url = "https://github.com/conan-io/conan-center-index"
-    description = "Library containing software that is broadly useful for image processing and image analysis applications."
+    description = "Library containing software that is broadly useful for " \
```
Comment (theirix) [inline recipes/leptonica/all/conanfile.py:11]: Great refactoring, thanks!

File: recipes/leptonica/all/conanfile.py [hunk 2]
```diff
@@ -50,106 +76,116 @@ def requirements(self):
...
+            tools.replace_in_file(cmakelists_src, "if (PNG_LIBRARIES)", "if(0)")
+            tools.replace_in_file(cmake_configure, "if (PNG_FOUND)", "if(0)")
+        if not self.options.with_tiff:
+            tools.replace_in_file(cmakelists_src, "if (TIFF_LIBRARIES)", "if(0)")
+            tools.replace_in_file(cmake_configure, "if (TIFF_FOUND)", "if(0)")
         if not self.options.with_openjpeg:
-            tools.replace_in_file(cmake_original,
-                                  "if(NOT JP2K)",
-                                  "if(FALSE)")
+            tools.replace_in_file(cmakelists_src, "if (JP2K_FOUND)", "if(0)")
```
Comment (theirix) [inline recipes/leptonica/all/conanfile.py:115]: Unfortunately, it is not enough for openjpeg/webp when pkgconfig is used.

Even when we unset `HAVE_LIBJP2K` in these branches and omit JP2K_INCLUDE_DIRS/JP2K_LIBRARIES from target, `HAVE_LIBJP2K` will be set in root `CMakeLists.txt` when pkgconfig succeeds with finding system openjpeg. One more patch is needed to replace `if(NOT JP2K)`. I can reproduce it only when a mentioned patch to `openjpeg-2.4`.

The same thing with WEBP.

File: recipes/leptonica/all/conanfile.py [hunk 3]
```diff
@@ -1,42 +1,68 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
 
     def requirements(self):
-        self.requires("zlib/1.2.11")
+        if self.options.with_zlib:
+            self.requires("zlib/1.2.11")
```
Comment (theirix) [inline recipes/leptonica/all/conanfile.py:65]: Comment about `self.requires("openjpeg/2.4.0")`:

Leptonica uses versioned paths (2.3, 2.4) in source files and cmake. To properly support openjpeg 2.4.0 we should replace `openjpeg-2.3` with `openjpeg-2.4`

File: recipes/leptonica/all/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/leptonica/all/conanfile.py:115]: Good catch, I'll fix that.

File: recipes/leptonica/all/conanfile.py (hunk 3 above)
Comment (SpaceIm) [reply recipes/leptonica/all/conanfile.py:65]: It doesn't use versioned paths if LIBJP2K_HEADER is not defined (which should be the case), so it should properly works.

Comment (theirix) [reply recipes/leptonica/all/conanfile.py:65]: That's right, it should work.

**Comment by SpaceIm** [comment]
I'm wondering if we shouldn't also force link to imported targets. Several dependencies have interface definitions which are not injected currently, maybe it breaks on Windows if some dependencies are shared (not tested).

**Comment by theirix** [comment]
> I'm wondering if we shouldn't also force link to imported targets. Several dependencies have interface definitions which are not injected currently, maybe it breaks on Windows if some dependencies are shared (not tested).

Usually, Windows builds with all-shared deps are problematic. CCI tests do not cover it, we can check it only manually.

**Review by theirix — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**